main

alexsolan1 opened the pull request.

alexsolan1 commented on main.cc:
```diff
@@ -0,0 +1,183 @@
+        m_data[m_len] = 0;
+    }
+    
+    String(const String& s) {
```
fdsfdsdsfds

iwrinc commented on main.cc:
```diff
@@ -0,0 +1,210 @@
+
+using namespace std;
+
+class String {
```
But how the user would use this class? What file should be included?

alexsolan1 replied:
I know I should separate the .h and the .cc, but I don't have a normal environment at home, so I work with the online compiler.

iwrinc commented on main.cc:
```diff
@@ -0,0 +1,210 @@
+#include <iostream>
+#include <string.h>
```
The standard says that whenever we need to include a C header xxx.h in C++, we have to `#include <cxxx>`

alexsolan1 replied:
Great. I didn't know. Now I know :) Fixed.

iwrinc commented on main.cc:
```diff
@@ -0,0 +1,210 @@
+    }
+    
+    String(unsigned capacity) {
+        m_data = new char[capacity];
```
Consider *initializing* the members, rather than assigning them.

alexsolan1 replied:
Ok, will do.

iwrinc commented on main.cc:
```diff
@@ -0,0 +1,210 @@
+    String(unsigned capacity) {
+        m_data = new char[capacity];
+        m_capacity = capacity;
+        m_length = 1;
```
Why 1? What does m_length denote?

alexsolan1 replied:
Right. Should be 0. Will fix.

iwrinc commented on main.cc:
```diff
@@ -0,0 +1,210 @@
+    }
+    
+    void reserve(unsigned capacity) {
+        if (capacity < m_capacity) { return; }
```
`<=`?

alexsolan1 replied:
Right, of course.

iwrinc commented on main.cc:
```diff
@@ -0,0 +1,210 @@
+        return m_capacity;
+    }
+
+    const char* data() const {
```
The assignment doesn't say that expicitly, but real std::string::data() allows data mutation.

iwrinc commented on main.cc:
```diff
@@ -0,0 +1,210 @@
+    
+    String& operator+=(const String& s) {
+        unsigned newlen = length() + s.length();
+        if (newlen + 1 >= m_capacity) {
```
`>`?

alexsolan1 replied:
Right again.

iwrinc commented on main.cc:
```diff
@@ -0,0 +1,210 @@
+        alloc_and_copy(data, length);
+    }
+    
+    String(unsigned capacity) {
```
Although it was not stated in the assignment explicitly, we'd like to mimic the standard container behavior. If a standard container's constructor get an integer, it is the desired *size* (not the capacity).   In order to set capacity, containers define `reserve` member function.

alexsolan1 replied:
But in this case, if it is the size and not the capacity, I am left with un-initialized data?

iwrinc replied:
You'll have _value-initialized_ data (please, read about value-initialization).

iwrinc commented on main.cc:
```diff
@@ -0,0 +1,210 @@
+    
+    void reallocate(unsigned length) {
+        String tmp(length);
+        tmp.m_length = m_length;
```
It looks like the above constructor is not "good enough", so you have to adjust some members after it. Note that after you change the semantics of the ctor, as described above, the things might look better here.

alexsolan1 replied:
But - I use the reallocate in the reserve, so I need it to adjust the capacity. Actually the naming here is not good. I will change it.

iwrinc commented on main.cc:
```diff
@@ -0,0 +1,210 @@
+        swap(*this, tmp);
+    }
+    
+    void add(const String& s, unsigned location) {
```
Very dangerous function. Consider calling it something like `unsafe_add`.

alexsolan1 replied:
Indeed (But it is private). I can change the name and add an assertion that capacity suffices.

iwrinc commented on main.cc:
```diff
@@ -0,0 +1,210 @@
+   
+private:
+    friend void swap(String& first, String& second) {
+        swap(first.m_capacity, second.m_capacity);
```
In this particular case all the members are built-in types. But what if some of them were a user-defined type with it's own `swap`?  
Note that `swap` is ADL-based. So the common idiom looks like this:
using std::swap;
swap(left, right);

alexsolan1 replied:
OK. Fixed.

iwrinc commented on main.cc:
```diff
@@ -0,0 +1,210 @@
+#include <string.h>
+#include <cassert>
+
+using namespace std;
```
Not in a namepsace/global scope, please!

alexsolan1 replied:
In the next exercises it is allowed... so I allowed myself in advance :) I promise that it just for this exercise...

iwrinc reviewed (commented):
Great! It looks much better now!

iwrinc commented on oneliners.cc:
```diff
@@ -0,0 +1,28 @@
+
+int main()
+{
+    vector<int> v(100);
+    iota(v.begin(), v.end(), 0);
```
But it's 2 lines (if I got it right, and it should be solution for (1) ).

alexsolan1 replied:
Hmm, right. Is something like 
vector<int> v{1,3,5,7,45} is an answer? or should we do something that can be extended to long vectors?

iwrinc replied:
Well, the assignment does not say you have push _different_ elements ;-)

alexsolan1 replied:
Yes, but if I should use it as a base for all the other operations, it will be a bad test case to have like
vector v{100, 5}...
Of course I could do it and then create a more interesting one, to test the other operations :)

iwrinc commented on one_liners2.cc:
```diff
@@ -0,0 +1,56 @@
+// 1. Xor to each char
+string xored_string(string s, char n) {
+    return accumulate(s.begin(), s.end(), string(""), [n](
+        const string& acc, const char& c){return acc + char(c^n);});
```
`const char& c` ?! :-)

iwrinc commented on one_liners2.cc:
```diff
@@ -0,0 +1,56 @@
+        const string& acc, const char& c){return acc + char(c^n);});
+}
+// reverse
+string my_reverse(string s) {
```
`const string&`

alexsolan1 replied:
Right

iwrinc commented on one_liners2.cc:
```diff
@@ -0,0 +1,56 @@
+// reverse
+string my_reverse(string s) {
+    return accumulate(s.crbegin(), s.crend(), string(""), [](
+        const string& acc, const char& c){return acc + c;});
```
Formally, that's a correct solution. But could you please think of a much more simple (and potentially fast) one?

alexsolan1 replied:
Actually, there is std::reverse, which is a one-liner and I am sure that it works faster than my one-liner, but it is probably not the purpose of this ex?

iwrinc replied:
I didn't mean this function, of course...
I guess the whole point of this exercise is to learn the rich (and some say bloated) interface of the standard library. So, just take a look at `std::string` interface :)

iwrinc commented on one_liners2.cc:
```diff
@@ -0,0 +1,56 @@
+bool all_capitals(const vector<string>& names) {
+    return accumulate(names.begin(), names.end(), true, [](
+        bool acc, const string& name){
+            return acc & 
```
You probably mean `&&`  (BTW, what would be the difference?)

alexsolan1 replied:
yeah, it is logical 'and' not bitwise.  Does the compiler converts the booleans to numeric to perform the  bitwise &?
Practically, I believe that although it is weird, it should work, as long as it is mapped true->1 and false->0.
I will fix, of course.

iwrinc replied:
Yes, in this case _integral promotion_ takes place (see C++20 standard, 7.3.6).  The problem is, however, that bitwise operations are _fully_ evaluated, while the logical ones have _short-circuit_ evaluation. So this is a nice example of how a single symbol might significantly affect performance :).

alexsolan1 replied:
Hmmm,  I see!

iwrinc commented on one_liners2.cc:
```diff
@@ -0,0 +1,56 @@
+vector<int> flat_vec(const vector<vector<int>>& vectors) {
+    return accumulate(vectors.begin(), vectors.end(), vector<int>{},
+        [](vector<int> acc, const vector<int>& cur) {
+            move(cur.begin(), cur.end(), back_inserter(acc));
+            return acc;
```
More than 1 line.

alexsolan1 replied:
It was specifically allowed to do 2 for this one.... :)

iwrinc replied:
oh, you're right!

iwrinc commented on optional.h:
```diff
@@ -0,0 +1,31 @@
+#include <string>
+
+template <class T>
+class Optional {
```
Please, test `Optional` with the following class:
```
struct my_class {
    my_class(int) {}
};
```

alexsolan1 replied:
It doesn't work since my_class is not std::is_default_constructible, I guess I should add the requirement to the definition of Optional.

alexsolan1 replied:
I can also implement my optional with variant, in this case it does not need to have T as default_constructible. But  I am not sure it is fair:)

alexsolan1 replied:
And I could also allocate memory on the stack for T, keep inside a pointer of T and when I need to create an object of T I will use this memory. Thus also  I won't need a default_constructible type.

iwrinc replied:
If the optional requires a default-constructible T, why couldn't we just use more simple and less obfuscated std::pair<bool, T>?
Of course it should not have such a constraint.
Besides, the assignment requires not to use free store ("heap").

alexsolan1 replied:
But using pair<bool, T> was exactly my naive implementation (just didn't use explicitly the pair)!
I understand that we want it work also with non default-constructible and therefore my solution is not good enough.
I didn't suggest to allocate memory on the heap, because the ex explicitly mentioned it.
I suggest to allocate memory on the stack (char buf[sizeof(T)])  and when we actually use the T call the new expression with this buffer
T* tptr = new(buf) T(t).

iwrinc commented on optional.h:
```diff
@@ -0,0 +1,31 @@
+		return m_valid;
+	}
+
+	T& get() {
```
So, what would happen if the user attempts to access "empty" optional (built in release mode)? What is the desired behavior?

alexsolan1 replied:
In release mode in my implementation he will have an empty object.
I could also throw an exception.
But I still have the feeling that I am doing something wrong here, since if T is a big object even when empty, we (I) allocate it on the stack even for non valid Optional.

iwrinc replied:
Please, read the assignment thoroughly.

alexsolan1 commented:
Something is weird with your last comments - I can't comment on each separately. But:
1) I added alignment
2) I added ctor with Forwarding (no additional copy)
3) I added throw exception (indeed it was written in the assignment and I missed it, sorry....)

iwrinc commented on optional.h:
```diff
@@ -12,8 +20,17 @@ class Optional {
 	Optional(const T& t) :
```
In what context is this ctor called now?

iwrinc commented on optional.h:
```diff
@@ -23,17 +40,21 @@ class Optional {
 		return *m_value;
 	}
 		
 	const T& get() const {
```
Cool! But now we have 2 identical functions (copy&paste). Could we do some code-reuse? (I.e., implement one function and call it from the other one)

iwrinc commented on optional.h:
```diff
@@ -23,17 +40,21 @@ class Optional {
 	
 private:
-	char buf[sizeof(T)];
+	alignas(alignof(T)) char buf[sizeof(T)];
```
Great! This is the right solution.
(Just note that we have `std::aligned_storage` exactly for this purpose.)

iwrinc commented on derive.h:
```diff
@@ -45,28 +49,32 @@ ostream& operator<<(ostream& os, const Expression& ex) {
 	{ }
 	
-	virtual bool is_zero() { return coef == 0.; }
+	virtual bool is_zero() { return coef() == 0.; }
```
Could it be `const`?

alexsolan1 replied:
Right. Fixed.

iwrinc commented on derive.h:
```diff
@@ -45,28 +49,32 @@ ostream& operator<<(ostream& os, const Expression& ex) {
+	double coef() const { return m_coef;}
+	friend class BinaryExpression;
+protected:
+	virtual optional<char> opt_var() const { return optional<char>{}; }
```
We have `std::nullopt` for this purpose.

alexsolan1 replied:
Ok, great, fixed.

iwrinc commented on derive.h:
```diff
@@ -45,28 +49,32 @@ ostream& operator<<(ostream& os, const Expression& ex) {
 
 class XExpression : public Expression {
 public:
 	virtual ostream& printTo(ostream& os) const {
```
Prefer `override`.

alexsolan1 replied:
OK.

iwrinc commented on derive.h:
```diff
@@ -45,28 +49,32 @@ ostream& operator<<(ostream& os, const Expression& ex) {
-	virtual bool is_zero() { return coef == 0.; }
+	virtual bool is_zero() { return coef() == 0.; }
 	virtual shared_ptr<Expression> derivative() const {
 		return std::dynamic_pointer_cast<Expression>(
```
Why is dynamic (or any) cast needed here?

alexsolan1 replied:
Indeed it seems that there is no  need.

iwrinc commented on derive.h:
```diff
@@ -0,0 +1,291 @@
+	);
+}
+
+shared_ptr<Expression> pow(shared_ptr<Expression> lhs, shared_ptr<Expression> rhs) {
```
Note that passing `shared_ptr` by value causes refcount increment, which involves expensive atomic operations.

alexsolan1 replied:
Fixed

iwrinc commented on derive.h:
```diff
@@ -0,0 +1,291 @@
+	);
+}
+
+shared_ptr<Expression> subtract(shared_ptr<Expression> lhs, shared_ptr<Expression> rhs) {
```
Looks like almost copy-paste from `add`, doesn't it?

alexsolan1 replied:
Of course. I thought to convert the -x to +(-1)*x, but I am not sure it worth it.

iwrinc commented on derive.cc:
```diff
@@ -0,0 +1,24 @@
+	cout << "h is: " << h << " derivative is: " << h.derivative() << endl;
+	cout << "f_mult_g is: " << f_mult_g << " derivative is: " << f_mult_g.derivative() << endl;
+	Func sub_func = f * 3. - x;
+	cout << "sub_func(x) = " << sub_func << " sub_func'(x) = " << sub_func.derivative() << endl;
```
Could you please add tests with zero coeff and very small coeff.

alexsolan1 replied:
I've got a test with zero coeff - line 18. I  will add now another test - more elaborate and a small coeff as well.

alexsolan1 replied:
Added another zero coeff and small coef.

iwrinc reviewed (commented):
General note: although this is legitimate design, it has some significant drawbacks  (I'd call it "javaish" design).
1) It uses enums and dynamic type queries, instead of C++ type system and compile-time checks. This leads to performance loss (dynamic casts are very slow, as they are usually implemented via string comparison) and poor maintainability (think of adding a new operation or expression type to all the numerous switch cases and dynamic type queries...).
2) It utilizes multiple virtual functions, which also might affect performance.

After you finish this solution, I'd recommend to try and propose an alternative design, which takes much more advantage of C++ compiler. Hint: templates and `boost::variant` :).
Such a solution would have its own advantages and drawbacks, and I believe it would be nice to compare them.

iwrinc commented on my_variant.h:
```diff
@@ -0,0 +1,118 @@
+ 		m_member(new (&m_aligned_storage) U{u} )
+	{ }
+	~my_variant() {
+		destroy_all<decltype(*this), T, types...>::f(*this);	
```
You solved the destructor problem "ad hoc", which is fine :). But actually, it was just a symptom of the more general problem: what if you (or the user!) had to apply some other function to the contained value (e.g. print() etc)? In your case the only such function was the destructor, that's why you encountered this problem here.
So, the general problem is the following: we need a *generic* mechanism that would allow us to apply any function to the *contained* type. When you have such a mechanism, you're able to invoke it in the destructor as well! (Hint: usually we call such a mechanism "visitor".)

alexsolan1 replied:
I added a visitor.

iwrinc commented on my_variant.h:
```diff
@@ -0,0 +1,118 @@
+	template <class U>
+	my_variant(U&& u) :
+		m_index(get_index<U, T, types...>::value),	
+ 		m_member(new (&m_aligned_storage) U{u} )
```
Looks like `&&` is unused...

alexsolan1 replied:
Fixed.

iwrinc commented on my_variant.h:
```diff
@@ -0,0 +1,118 @@
+
+template <class Obj, class... types> struct destroy_all;
+template <class T, class... types>
+struct my_variant {
```
Missing assignment (copy/move)

alexsolan1 replied:
Yes, but it is more technical now and we were not requested to write  a full solution there.

iwrinc commented on my_variant.h:
```diff
@@ -0,0 +1,118 @@
+    std::integral_constant<std::size_t, 1 + get_index<T, Ts...>::value> {};
+
+template <typename T>
+struct get_index<T> {
```
This probably should have been the implementation of the primary template (line 41)? Anyway, static assert here disables any attempt of SFINAE. Is this intentional?

alexsolan1 replied:
So, I am actually not sure what is the right interface. What do we expect for 
get_index<MyClass, double, string, MyOtherClass>::value?

alexsolan1 replied:
Ok, I understand now. Maybe it is better just to remove it and indeed leave option for SFINAE. Will do.

iwrinc commented on my_variant.h:
```diff
@@ -0,0 +1,118 @@
+	// need to add emplacement ctors...
+
+	template <class U>
+	my_variant(U&& u) :
```
It seems that you disallow the following, don't you: 
`my_variant<size_t> v(1); // please add such a test `

alexsolan1 replied:
Hmmm, you are right. Will fix it.

alexsolan1 replied:
Yes, I forget the Head in the max_size, max_align calcs, but now it works.

iwrinc commented on my_variant.h:
```diff
@@ -0,0 +1,118 @@
+};
+
+template <class U, class... types>
+U& get(my_variant<types...>& my) {
```
All these templates probably should not be a part of the variant's namespace/interface.

alexsolan1 replied:
Right, but since we are not required to do  a full solution here, I will leave it be.

iwrinc commented on my_variant.h:
```diff
@@ -0,0 +1,118 @@
+template <typename T, typename... Ts> struct get_index;
+
+template <typename T, typename... Ts>
+struct get_index<T, T, Ts...> : std::integral_constant<std::size_t, 0> {};
```
There is some inconsistency: here you use `integral_constant` (which is fine), and above helpers - manual implementation.

alexsolan1 replied:
Since it is an exercise, I played with 2 different styles of doing it. You are right that if it was a production code, the inconsistency is confusing.

RotemKalisch commented on derive.h:
```diff
@@ -0,0 +1,311 @@
+	return os;
+}
+
+class Expression {
```
Base class with virtual functions should always have a virtual destructor
```
virtual ~Expression() {}
```

alexsolan1 replied:
You are right. Added.

RotemKalisch commented on derive.h:
```diff
@@ -0,0 +1,311 @@
+					multiply(rhs->derivative(), lhs)
+				);
+			case power: 
+				{
```
Why is this new block?

alexsolan1 replied:
Because I am using a variable inside and the compiler does not agree to it if it is not in a separate block.

RotemKalisch commented on derive.h:
```diff
@@ -0,0 +1,311 @@
+
+class Func {
+public:
+	Func(double coef) :
```
Convention is:```
Func(...) :
    ....
    ....
{
```

RotemKalisch commented on derive.h:
```diff
@@ -0,0 +1,311 @@
+	{ }	
+
+	Func(Func&& func) : 
+		m_expr(func.m_expr) {
```
```
m_expr(std::move(func.m_expr))
```
also then there's no need to do func.m_expr = nullptr

alexsolan1 replied:
Right! Fixed.

RotemKalisch commented on my_variant.h:
```diff
@@ -0,0 +1,149 @@
+	my_variant() : m_member(new (&m_aligned_storage) Head{}) { }
+
+	my_variant(my_variant<Head, types...>&) {
+		assert(sizeof(Head) == 0 && "Not implemented");
```
Are you using and operator on a string? What am I missing?

alexsolan1 replied:
I am using a limited c assert which does not provide an interface for a message (like static_assert) and this is the patchy way to issue a message. If the lhs condition is not satisfied the rhs is ignored, otherwise the string will be part of the message. It is ugly, but in real life I will have nice library Macros/Methods to it :)
